### PR TITLE
Explicit cast should warn if nested nullability is not explicitly convertible

### LINKED ISSUE #29641: Explicit cast should warn if nested nullability does not match
The second conversion, from `List<string>` to `List<string?>` should result in a warning.
```c#
using System.Collections.Generic;
using System.Runtime.CompilerServices;
[NonNullTypes]
class C
{
    static void F(List<string> x)
    {
        var y = (List<object>)x;  // error CS0030: Cannot convert 'List<string>' to 'List<object>'
        var z = (List<string?>)x; // no warning
    }
}
```

### PR DESCRIPTION
**Pull request by cston**
Fixes #29641

### CODE REVIEW

**Comment by jcouv** [comment]
Hum, CI didn't seem to trigger properly...

File: docs/features/nullable-reference-types.md [hunk 1]
```diff
@@ -104,13 +104,28 @@ string t = default; // assigns ?, warning
...
 
-### Conversions
+### Implicit conversions
 _Describe valid top-level and variance conversions._
 _Describe warnings from user-defined conversions._
 
+### Explicit conversions
+The top-level nullability of `(X)y` is the nullability of `y`.
+The top-level nullability of `(X?)y` is `?`.
+A W warning is reported is reported when converting `?` to `!`.
```
Comment (jcouv) [inline docs/features/nullable-reference-types.md]: >is reported [](start = 24, length = 11)

duplicate

File: docs/features/nullable-reference-types.md [hunk 2]
```diff
@@ -104,13 +104,28 @@ string t = default; // assigns ?, warning
...
+```c#
+var s = (string)default(string); // x is string?, warning converting null
```
Comment (jcouv) [inline docs/features/nullable-reference-types.md]: >x [](start = 36, length = 1)

wrong local name. Also on next line

File: docs/features/nullable-reference-types.md [hunk 3]
```diff
@@ -104,13 +104,28 @@ string t = default; // assigns ?, warning
 T t = default; // assigns ?, warning
 ```
 
-### Conversions
+### Implicit conversions
 _Describe valid top-level and variance conversions._
 _Describe warnings from user-defined conversions._
 
+### Explicit conversions
```
Comment (jcouv) [inline docs/features/nullable-reference-types.md]: Should merge with section on casts below.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -29381,9 +29384,21 @@ static void F(C<object>? x, C<object?> y)
             var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
             // PROTOTYPE(NullableReferenceTypes): Should there be a warning on c1 too?
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >// PROTOTYPE(NullableReferenceTypes): Should there be a warning on c1 too? [](start = 12, length = 74)

PROTOTYPE can be removed?

File: docs/features/nullable-reference-types.md [hunk 5]
```diff
@@ -152,11 +152,11 @@ var x = (string)maybeNull; // x is string?, W warning
...
 ```
-Explicit casts change nested nullability.
-_Should a warning be reported when there is not an implicit nullable conversion (for instance `(List<object>)new List<object?>()`)?_
+A warning is reported if the type of operand including nested nullability is not explicitly convertible to the target type.
 ```c#
-List<object?> x = ...;
-var y = (List<object>)x; // y is List<object!>!, no warning
+var x =  new List<string>();
+var y = (List<string?>)x;        // warning
+var z = (IEnumerable<string?>)x; // no warning
```
Comment (AlekseyTs) [inline docs/features/nullable-reference-types.md]: Is there a test for this scenario?

Comment (cston) [reply docs/features/nullable-reference-types.md]: Good catch. Added.

---
In reply to: [215080231](https://github.com/dotnet/roslyn/pull/29665#discussion_r215080231) [](ancestors = 215080231)

**Pull request merged**